### PR TITLE
C++: include stack-allocated arrays in off-by-one query

### PR DESCRIPTION
**Pull request by rdmarsh2**
Depends on #13045. Only `a96cc6d` is new.

### CODE REVIEW

File: cpp/ql/src/experimental/Security/CWE/CWE-193/ConstantSizeArrayOffByOne.ql [hunk 1]
```diff
@@ -79,23 +79,28 @@ predicate isInvalidPointerDerefSink2(DataFlow::Node sink, Instruction i, string
...
+  v.getUnspecifiedType().(ArrayType).getArraySize() = size and
   semBounded(getSemanticExpr(pai.getRight()), any(SemZeroBound b), bound, true, _) and
   delta = bound - size
 }
 
-module FieldAddressToDerefConfig implements DataFlow::StateConfigSig {
+module ArrayAddressToDerefConfig implements DataFlow::StateConfigSig {
   newtype FlowState =
-    additional TArray(Field f) or
+    additional TArray(Variable v) or
```
Comment (MathiasVP) [inline cpp/ql/src/experimental/Security/CWE/CWE-193/ConstantSizeArrayOffByOne.ql:92]: As I mentioned in https://github.com/github/codeql/pull/13045, we can probably restrict these variables to be the ones satisfying `pointerArithOverflow(_, v, _, _, _)`.

**Review by MathiasVP — COMMENTED** [review]
a96cc6d4da425bbf88d91ba8c6bf5faf89a64ca0 looks good to me (with the exception of a small comment which will likely be resolved in the base PR anyway).

**Review by MathiasVP — COMMENTED** [review]
This LGTM once:
- We have a happy DCA run, and
- We've addressed the two `Omittable 'exists' variable` alerts. I think they highlight the possibility of a nice cleanup.

File: cpp/ql/src/experimental/Security/CWE/CWE-193/ConstantSizeArrayOffByOne.ql [hunk 2]
```diff
@@ -78,28 +78,39 @@
...
+bindingset[pai, size]
+predicate pointerArithOverflow0Impl(PointerArithmeticInstruction pai, int size, int bound, int delta) {
+  constantUpperBounded(pai, bound) and
   delta = bound - size and
   delta >= 0 and
   size != 0 and
   size != 1
 }
 
+predicate pointerArithOverflow0(PointerArithmeticInstruction pai, int delta) {
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/experimental/Security/CWE/CWE-193/ConstantSizeArrayOffByOne.ql:111]: Fixed in 3b4f2b22d65dc9de7b7df140b043243bd87bee6b.

File: cpp/ql/src/experimental/Security/CWE/CWE-193/ConstantSizeArrayOffByOne.ql [hunk 3]
```diff
@@ -78,28 +78,39 @@
...
+  exists(int size, int bound |
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/experimental/Security/CWE/CWE-193/ConstantSizeArrayOffByOne.ql]: Fixed in 3b4f2b22d65dc9de7b7df140b043243bd87bee6b.

**Comment by MathiasVP** [comment]
I'm looking at results now, and I'll update this message as I go through them.

### Samate

All the new results are classified as TPs in Samate 🎉.

**Review by MathiasVP — APPROVED** [review]
I'll approve this so that @rdmarsh2 can merge the PR once he's looked at my changes (i.e., e32f7d84a5303da643fe4eb398ad2568cb57949c and onwards) and the analysis I did of the DCA results.

**Comment by rdmarsh2** [comment]
I'm a little surprised [e32f7d8](https://github.com/github/codeql/pull/13116/commits/e32f7d84a5303da643fe4eb398ad2568cb57949c) is a performance win, but DCA is definitely happier afterwards. LGTM

**Pull request merged**